### PR TITLE
Make loaded datasets similar to new ones

### LINKED ISSUE #1106: dataset.completed does not work
### Steps to reproduce

```
dataset = load_by_id(1)
dataset.completed
```

Output:

> Traceback (most recent call last):
> 
>   File "<ipython-input-178-f869a978e653>", line 1, in <module>
>     dataset.completed
> 
>   File "c:\users\triton5acq\qcodes\qcodes\dataset\data_set.py", line 306, in completed
>     return self._completed
> 
> AttributeError: 'DataSet' object has no attribute '_completed'

### System
**Windows 7**

**qcodes/master**

**19a2016f9c2d999027b3949a646200d2d5b41e7e**

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
Fixes #1106

Changes proposed in this pull request:
- Tie the `DataSet` object somewhat stronger to the sqlite DB
- Make loaded `DataSet`s have the same attributes as newly created ones
- Add a few tests

Also clean up a little in the underlying sql functions (see the `mark_run` function).

@jenshnielsen @sohailc

### CODE REVIEW

File: qcodes/dataset/data_set.py [hunk 1]
```diff
@@ -658,16 +677,27 @@ def new_data_set(name, exp_id: Optional[int] = None,
...
-        if len(get_experiments(d.conn)) > 0:
-            exp_id = get_last_experiment(d.conn)
+        if len(get_experiments(conn)) > 0:
+            exp_id = get_last_experiment(conn)
         else:
             raise ValueError("No experiments found."
                              "You can start a new one with:"
                              " new_experiment(name, sample_name)")
+    # This is admittedly a bit weird. We create a dataset, link it to some
+    # run in the DB and then (using _new) change what it's linked to
```
Comment (sohailc) [inline qcodes/dataset/data_set.py]: Can you also explain why you are doing weird stuff like that?

**Comment by sohailc** [comment]
Your comment "Make loaded DataSets have the same attributes as newly created ones" makes me a bit nervous. A good design principle is that classes should define all attributes in __init__. This way all instances of a class will always have the same attributes. Let's make sure that this convention is followed everywhere.

**Comment by sohailc** [comment]
In fact, come to think of it, the title "Make loaded datasets similar to new ones" makes me really nervous! What do you mean, make it 'similar' to new ones?! Are loaded and new data sets not simply instances of the same class?

**Comment by sohailc** [comment]
Thanks for your comments, William. I will look into 1106 and think about a good solution

**Comment by WilliamHPNielsen** [comment]
@QCoDeS/core while waiting for a "deeper" solution, perhaps we can accept this quickfix that will allow users to load datasets and see whether they are completed?

**Comment by jenshnielsen** [comment]
The best alternative solution that @Dominik-Vogel and I can see is to modify the init method so it can do the work of all the new functions. That admittedly results in a somewhat convoluted init function but perhaps it's still better

**Comment by WilliamHPNielsen** [comment]
I think either way is fine, but I think we should be reasonably quick in applying the fix.

**Comment by WilliamHPNielsen** [comment]
In the meantime, let me just cross-link to #990, since that actually taps in to the same problem.

**Comment by WilliamHPNielsen** [comment]
@astafan8 I think that is a good way to proceed. The AppVeyor business is not directly related to this PR, but feel free to extend the deadlines and push that to this branch if you want. Else I think we can just merge?

**Comment by astafan8** [comment]
@WilliamHPNielsen i had a quick look at the AppVeyor output. I think indeed we can merge. Shall i do it?

**Pull request merged**